library/alloc/src/boxed/thin.rs

```diff
@@ -91,6 +95,65 @@ impl<T> ThinBox<T> {
 
 #[unstable(feature = "thin_box", issue = "92791")]
 impl<Dyn: ?Sized> ThinBox<Dyn> {
+    #[cfg(not(no_global_oom_handling))]
+    fn new_unsize_zst<T>(value: T) -> Self
+    where
+        T: Unsize<Dyn>,
+    {
+        assert!(mem::size_of::<T>() == 0);
+
+        const fn max(a: usize, b: usize) -> usize {
+            if a > b { a } else { b }
+        }
+
+        // We are using `[T; 0]` here instead of `T` because
+        // compiler instantiates this code even for non-ZST,
+        // and reports a pointer going beyond the allocation.
+        let alloc: &[T; 0] = const {
+            // ZST T pointer must be aligned to both the value and the metadata.
+            // Metadata must be at the address `&T - size_of::<Metadata>`.
+            // This is how we need to allocate the memory:
+            // [ padding | metadata | ZST T value ]
+
+            // FIXME: just call `WithHeader::alloc_layout` with size reset to 0.
+            // Currently that's blocked on `Layout::extend` not being `const fn`.
+
+            let alloc_align =
+                max(mem::align_of::<T>(), mem::align_of::<<Dyn as Pointee>::Metadata>());
+
+            let alloc_size =
+                max(mem::align_of::<T>(), mem::size_of::<<Dyn as Pointee>::Metadata>());
+
+            unsafe {
+                // SAFETY: align is power of two because it is the maximum of two alignments.
+                let alloc: *mut u8 = const_allocate(alloc_size, alloc_align);
+
+                // Zerofill to be safe.
+                // SAFETY: `alloc_size` is the size of the allocation.
+                ptr::write_bytes(alloc, 0, alloc_size);
+
+                // SAFETY: adding offset within the allocation.
+                let metadata_ptr: *mut u8 = alloc.add(
+                    alloc_size.checked_sub(mem::size_of::<<Dyn as Pointee>::Metadata>()).unwrap(),
+                );
```

Comment:
Let's just always use pointer methods.
```suggestion
                let metadata_ptr: *mut u8 = alloc.add(alloc_size).cast::<<Dyn as Pointee>::Metadata>().sub(1);
```

Reply:
Done.